<a href="https://colab.research.google.com/github/pingao2019/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/PA2_Copy_of_LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [0]:
import pandas as pd
df=pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data', header=None)
print(df.shape)
df.head()


(435, 17)


0  1  2  3  4  5  6  7  8  9  10 11 12 13 14 15 16
0  republican  n  y  n  y  y  y  n  n  n  y  ?  y  y  y  n  y
1  republican  n  y  n  y  y  y  n  n  n  n  n  y  y  y  n  ?
2    democrat  ?  y  y  ?  y  y  n  n  n  n  y  n  y  y  n  n
3    democrat  n  y  y  n  ?  y  n  n  n  n  y  n  y  n  n  y
4    democrat  y  y  y  n  y  y  n  n  n  n  y  ?  y  y  y  y

In [0]:
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data',
                 names=['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime',
                          'duty-free', 'south-africa'],
                 na_values="?")
print(df.shape)
df.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n          NaN
2    democrat                 NaN             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [0]:
# One way to replace question marks with NaNs
# df = df.replace({"?": np.NaN})

In [0]:
df= df.replace({'y':1,'n':0})
df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [0]:
df['party'].value_counts()

democrat      267
republican    168
Name: party, dtype: int64

In [0]:
rep=df[df['party']=='republican']
rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [0]:
dem=df[df['party']=='democrat']
dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

In [0]:
from scipy.stats import ttest_ind, ttest_ind_from_stats,ttest_rel


water project bill


1) Null hypothesis: there is no different between average voting rates for the "water project" bill between replican and demcrat.



$\bar{x}_{1}==\bar{x}_{2}$            ($\bar{x}$ is the mean of votes)

2) Alternatives Hypothesis: $\bar{x}_{1} \neg \bar{x}_{2}$ not equal.
   This means level support between the two parties.

3).95% Confidence Level

In [0]:
rep['water-project'].mean()

0.5067567567567568

In [0]:
dem['water-project'].mean()

0.502092050209205

In [0]:
from scipy.stats import ttest_ind

ttest_ind(rep['water-project'], dem['water-project'],nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

In [0]:
col= rep['water-project']
rep_water_project_no_nans=col[-np.isnan(col)]
col= dem['water-project']
dem_water_project_no_nans=col[-np.isnan(col)]
print(len(rep_water_project_no_nans))
print(len(dem_water_project_no_nans))

148
239


degree of freedom df= 148-1=147
4) T- statistics: .089
5) P-value: .929

it fails to reject null hypothesis because p-value>0.05. ( it not mean accept null hypothesis.( mean of rep=0.5067567567567568, dem=0.502092050209205. )





       The one sample t-test.


 let look dem support "south-africa"bill

1a) Null hypothesis:
$\bar{x}_{1}$(average dem support for SA bill ==1.
This says 100% of dem support the bill. means the party is full support.
1b) Null hypothesis:
$\bar{x}_{1}$(average dem support for SA bill ==.5
This says 50% of dem support the bill. means the party is split.

1c) Null hypothesis:
$\bar{x}_{1}$(average dem support for SA bill ==0.
This says 0% of dem support the bill. means the party all not support.


1a) Null hypothesis:
$\bar{x}_{1}$(average dem support for SA bill ==.78245
This says  0% of dem support the bill. means the party is full support.



** with 1-sample_t-test we can frame taht I am asking through my choice of null hypothesis

1) Null hypothesis:
$\bar{x}_{1}$ average dem support for SA bill ==.5
This says 50% of dem support the bill. means the party is split.
2) altenative hypothesis: Support is not eaqual to .5 or 50%
$\bar{x}_{1}$ not equal .5

This say nothig about if support is greater than or less than 50%. It is just saying that it is 50%-it is something other than 50%.



In [0]:
from scipy.stats import ttest_1samp
ttest_1samp(dem['south-africa'], .5, nan_policy='omit')

Ttest_1sampResult(statistic=23.965755112488182, pvalue=1.7513759267849718e-58)

p-value is so small, we reject null hypothesis(null hypothesis support for the bill is .5(split party)). conclude that something is different.

In [0]:
dem['south-africa'].mean()

0.9351351351351351

In [0]:
ttest_1samp(dem['south-africa'], .9, nan_policy='omit')

Ttest_1sampResult(statistic=1.935123083617057, pvalue=0.05450836316336539)

p-value =.0545<0.05,  so fail to reject the null hypothesis. the dem for supporting the bill is not significantlty different from 90%

In [0]:
rep.columns.to_list()

['party',
 'handicapped-infants',
 'water-project',
 'budget',
 'physician-fee-freeze',
 'el-salvador-aid',
 'religious-groups',
 'anti-satellite-ban',
 'aid-to-contras',
 'mx-missile',
 'immigration',
 'synfuels',
 'education',
 'right-to-sue',
 'crime',
 'duty-free',
 'south-africa']

In [0]:
rep['budget'].mean()

0.13414634146341464

In [0]:
dem['budget'].mean()

0.8884615384615384

In [0]:
# budget bill
col= rep['budget']
rep_budget_no_nans=col[-np.isnan(col)]
col= dem['budget']
dem_budget_no_nans=col[-np.isnan(col)]
print(len(rep_budget_no_nans))
print(len(dem_budget_no_nans))

ttest_ind(rep['budget'], dem['budget'],nan_policy='omit')


164
260


Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77)

# using 2 samples t-test by 1). Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
#2). Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
#3).Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

In [0]:
  # immigration
col= rep['immigration']
rep_immigration_no_nans=col[-np.isnan(col)]
col= dem['immigration']
dem_immigration_no_nans=col[-np.isnan(col)]
print(len(rep_immigration_no_nans))
print(len(dem_immigration_no_nans))
ttest_ind(rep['immigration'], dem['immigration'],nan_policy='omit')

165
263


Ttest_indResult(statistic=1.7359117329695164, pvalue=0.08330248490425066)

In [0]:
rep_immigration_no_nans.mean()

0.5575757575757576

In [0]:
dem_immigration_no_nans.mean()

0.4714828897338403

1) Null hypothesis : there is no difference between avaerage voting rates between two parties on immigration bill.                                         
2) p-value = 0.08330248490425066> 0.01, so it fail to reject null hypothesis,
   so there is no big difference between two parties on immigration bill. 

In [0]:
  #  education
col= rep['education']
rep_education_no_nans=col[-np.isnan(col)]
col= dem['education']
dem_education_no_nans=col[-np.isnan(col)]
print(len(rep_education_no_nans))
print(len(dem_education_no_nans))
ttest_ind(rep['education'], dem['education'],nan_policy='omit')

155
249


Ttest_indResult(statistic=20.500685724563073, pvalue=1.8834203990450192e-64)

In [0]:
  #  crime
col= rep['crime']
rep_crime_no_nans=col[-np.isnan(col)]
col= dem['crime']
dem_crime_no_nans=col[-np.isnan(col)]
print(len(rep_crime_no_nans))
print(len(dem_crime_no_nans))
ttest_ind(rep['crime'], dem['crime'],nan_policy='omit')

161
257


Ttest_indResult(statistic=16.342085656197696, pvalue=9.952342705606092e-47)

In [0]:
  #  synfuels bill
col= rep['synfuels']
rep_synfuels_no_nans=col[-np.isnan(col)]
col= dem['synfuels']
dem_synfuels_no_nans=col[-np.isnan(col)]

ttest_ind(rep['synfuels'], dem['synfuels'],nan_policy='omit')

Ttest_indResult(statistic=-8.293603989407588, pvalue=1.5759322301054064e-15)

In [0]:
  #  right-to-sue bill
col= rep['right-to-sue']
rep_right_to_sue_no_nans=col[-np.isnan(col)]
col= dem['right-to-sue']
dem_right_to_sue_no_nans=col[-np.isnan(col)]

ttest_ind(rep['right-to-sue'], dem['right-to-sue'],nan_policy='omit')

Ttest_indResult(statistic=13.51064251060933, pvalue=1.2278581709672758e-34)

In [0]:
  #  religious-groups bill
col= rep['religious-groups']
rep_religious_groups_no_nans=col[-np.isnan(col)]
col= dem['right-to-sue']
dem_religious_groups_no_nans=col[-np.isnan(col)]

ttest_ind(rep['religious-groups'], dem['religious-groups'],nan_policy='omit')

Ttest_indResult(statistic=9.737575825219457, pvalue=2.3936722520597287e-20)

In [0]:
rep_religious_groups_no_nans.mean()

0.8975903614457831

In [0]:
dem_religious_groups_no_nans.mean()

0.2896825396825397

1) Null hypothesis : there is no difference between avaerage voting rates between two parties on religious-groups  bill.                                         
2) p-value =< 0.01, so it   rejects null hypothesis,
   so there is  difference between two parties on religious-groups bill.
3) mean of republican voting is more than demcrats, so republicans support more than democrats with p < 0.01

In [0]:
  #  mx-missile bill
col= rep['mx-missile']
rep_mx_missile_no_nans=col[-np.isnan(col)]
col= dem['mx-missile']
dem_mx_missile_no_nans=col[-np.isnan(col)]

ttest_ind(rep['mx-missile'], dem['mx-missile'],nan_policy='omit')

Ttest_indResult(statistic=-16.437503268542994, pvalue=5.03079265310811e-47)

In [0]:
rep_mx_missile_no_nans.mean()

0.11515151515151516

In [0]:
dem_mx_missile_no_nans.mean()

0.7580645161290323

1) Null hypothesis : there is no difference between avaerage voting rates between two parties on mx-missile  bill.                                         
2) p-value =< 0.01, so it   rejects null hypothesis,
   so there is  difference between two parties on mx-missile bill.
3) mean of republican voting is less than demcrats, so democrats support more than republicans with p < 0.01